File: question/leetcode/leetcode0238.Product-of-Array-Except-Self/leetcode0238.go
package leetcode

func productExceptSelf(nums []int) []int {
	ans := make([]int, len(nums))
	tmp := 1

	// 初始化ans数组
	for i := range ans {
		ans[i] = 1
	}

	for i := 1; i < len(nums); i++ {
		ans[i] = ans[i-1] * nums[i-1]
	}

	for i := len(nums) - 2; i >= 0; i-- {
		tmp *= nums[i+1]
		ans[i] *= tmp
	}

	return ans
}

File: question/leetcode/leetcode0238.Product-of-Array-Except-Self/leetcode0238.py
from typing import List


def productExceptSelf(nums: List[int]) -> List[int]:
    n = len(nums)
    ans = [1] * n

    # 左边所有数的乘积
    for i in range(1, n):
        ans[i] = ans[i - 1] * nums[i - 1]

    # 右边所有数的乘积
    right = 1
    for i in range(n - 2, -1, -1):
        right *= nums[i + 1]
        ans[i] *= right

    return ans
